swap/protocol.go

```diff
@@ -63,50 +66,70 @@ func (s *Swap) APIs() []rpc.API {
 	}
 }
 
+// Start is a node.Service interface method
 func (s *Swap) Start(server *p2p.Server) error {
 	log.Info("Swap service started")
 	return nil
 }
 
+// Stop is a node.Service interface method
 func (s *Swap) Stop() error {
 	return nil
 }
 
 // verifyHandshake verifies the chequebook address transmitted in the swap handshake
 func (s *Swap) verifyHandshake(msg interface{}) error {
-	handshake, ok := msg.(*SwapHandshakeMsg)
+	handshake, ok := msg.(*HandshakeMsg)
 	if !ok || (handshake.ContractAddress == common.Address{}) {
 		return ErrEmptyAddressInSignature
 	}
 
 	return s.verifyContract(context.TODO(), handshake.ContractAddress)
 }
 
+// run is the actual swap protocol run method
 func (s *Swap) run(p *p2p.Peer, rw p2p.MsgReadWriter) error {
 	protoPeer := protocols.NewPeer(p, rw, Spec)
 
-	answer, err := protoPeer.Handshake(context.TODO(), &SwapHandshakeMsg{
+	answer, err := protoPeer.Handshake(context.TODO(), &HandshakeMsg{
 		ContractAddress: s.owner.Contract,
 	}, s.verifyHandshake)
 	if err != nil {
 		return err
 	}
 
-	beneficiary, err := s.getContractOwner(context.TODO(), answer.(*SwapHandshakeMsg).ContractAddress)
+	beneficiary, err := s.getContractOwner(context.TODO(), answer.(*HandshakeMsg).ContractAddress)
 	if err != nil {
 		return err
 	}
 
-	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, answer.(*SwapHandshakeMsg).ContractAddress)
-
-	s.lock.Lock()
-	s.peers[p.ID()] = swapPeer
-	s.lock.Unlock()
+	swapPeer := NewPeer(protoPeer, s, s.backend, beneficiary, answer.(*HandshakeMsg).ContractAddress)
+	s.addPeer(swapPeer)
+	defer s.removePeer(swapPeer)
```

Comment:
Why do we remove the peer once the inner forever loop has stopped? Is it due to resources?

Reply:
Yes, mainly. Check https://github.com/ethersphere/swarm/pull/1554#discussion_r304513162

Reply:
Got it, thanks for the clarification.